adapters/repos/db/lsmkv/store.go

```diff
@@ -115,17 +117,43 @@ func (s *Store) bucketDir(bucketName string) string {
 func (s *Store) CreateOrLoadBucket(ctx context.Context, bucketName string,
 	opts ...BucketOption,
 ) error {
-	if b := s.Bucket(bucketName); b != nil {
+	var bucketLock *sync.Mutex
+
+	s.bucketAccessLock.Lock()
```

Comment:
**suggestion:**  i see multiple `s.bucketAccessLock`  lock/unlock , why not  just  having  it at the beginning of the  function 
``` go
s.bucketAccessLock.Lock()
defer s.bucketAccessLock.UnLock()
```

Reply:
why not doing something like that ? 

```go
func (s *Store) CreateOrLoadBucket(ctx context.Context, bucketName string,
	opts ...BucketOption,
) error {
	s.bucketAccessLock.Lock()
	defer s.bucketAccessLock.Unlock()

	if _, bucketExists := s.bucketsByName[bucketName]; bucketExists {
		return nil
	}

	b, err := NewBucket(ctx, s.bucketDir(bucketName), s.rootDir, s.logger, s.metrics,
		s.cycleCallbacks.compactionCallbacks, s.cycleCallbacks.flushCallbacks, opts...)
	if err != nil {
		return err
	}

	s.bucketsByName[bucketName] = b
	return nil
}
```
which originally  was 

```go
func (s *Store) CreateOrLoadBucket(ctx context.Context, bucketName string,
	opts ...BucketOption,
) error {
       // this thread safe
	if b := s.Bucket(bucketName); b != nil {
		return nil
	}

      // having lock here is the only diff 
	b, err := NewBucket(ctx, s.bucketDir(bucketName), s.rootDir, s.logger, s.metrics,
		s.cycleCallbacks.compactionCallbacks, s.cycleCallbacks.flushCallbacks, opts...)
	if err != nil {
		return err
	}

       // this thread safe
	s.setBucket(bucketName, b)
	return nil
}
```

Reply:
> **suggestion:** i see multiple `s.bucketAccessLock` lock/unlock , why not just having it at the beginning of the function
> 
> ```go
> s.bucketAccessLock.Lock()
> defer s.bucketAccessLock.UnLock()
> ```

if the bucketAccessLock lock is hold during bucket initialization (which could take considerable time), it won't be possible to concurrently load another buckets.

Functions like `setBucket` can still be used

Reply:
gotcha, but then isn't the original already thread safe ? or `NewBucket()` tries to access `bucketsByName` ?

Reply:
> gotcha, but then isn't the original already thread safe ? or `NewBucket()` tries to access `bucketsByName` ?

with the implementation without this changes it was possible to initiate the same bucket multiple times, which was the reason of inconsistencies being generated

Reply:
we caught up on it and we will revisit it later on in a separate PR